src/coreclr/jit/hwintrinsicarm64.cpp

```diff
@@ -210,6 +210,140 @@ bool HWIntrinsicInfo::isScalarIsa(CORINFO_InstructionSet isa)
+//
+// Arguments:
+//    intrinsic            -- NamedIntrinsic associated with the HWIntrinsic to lookup
+//    sig                  -- signature of the intrinsic call.
+//    immNumber            -- Which immediate to use (1 for most intrinsics)
+//    simdBaseType         -- base type of the intrinsic
+//    simdType             -- vector size of the intrinsic
+//    op2ClsHnd            -- cls handler for op2
+//    op2ClsHnd            -- cls handler for op3
+//    immSimdSize [OUT]    -- Size of the immediate
+//    pImmUpperBound [OUT] -- Base type of the immediate
+//
+void Compiler::getHWIntrinsicImmTypes(NamedIntrinsic       intrinsic,
+                                      CORINFO_SIG_INFO*    sig,
+                                      unsigned             immNumber,
+                                      var_types            simdBaseType,
+                                      CorInfoType          simdBaseJitType,
+                                      CORINFO_CLASS_HANDLE op2ClsHnd,
+                                      CORINFO_CLASS_HANDLE op3ClsHnd,
+                                      unsigned*            immSimdSize,
+                                      var_types*           immSimdBaseType)
+{
+    HWIntrinsicCategory category = HWIntrinsicInfo::lookupCategory(intrinsic);
+
+    if (category == HW_Category_SIMDByIndexedElement)
+    {
+        assert(immNumber == 1);
+
+        CorInfoType indexedElementBaseJitType;
+        var_types   indexedElementBaseType;
+        *immSimdSize = 0;
+
+        if (sig->numArgs == 3)
+        {
+            indexedElementBaseJitType = getBaseJitTypeAndSizeOfSIMDType(op2ClsHnd, immSimdSize);
+            indexedElementBaseType    = JitType2PreciseVarType(indexedElementBaseJitType);
+        }
+        else
+        {
+            assert(sig->numArgs == 4);
+            indexedElementBaseJitType = getBaseJitTypeAndSizeOfSIMDType(op3ClsHnd, immSimdSize);
+            indexedElementBaseType    = JitType2PreciseVarType(indexedElementBaseJitType);
+
+            if (intrinsic == NI_Dp_DotProductBySelectedQuadruplet)
+            {
+                assert(((simdBaseType == TYP_INT) && (indexedElementBaseType == TYP_BYTE)) ||
+                       ((simdBaseType == TYP_UINT) && (indexedElementBaseType == TYP_UBYTE)));
+                // The second source operand of sdot, udot instructions is an indexed 32-bit element.
+                indexedElementBaseType = simdBaseType;
+            }
+        }
+
+        assert(indexedElementBaseType == simdBaseType);
+    }
+    else if (intrinsic == NI_AdvSimd_Arm64_InsertSelectedScalar)
+    {
+        if (immNumber == 2)
+        {
+            CorInfoType otherBaseJitType = getBaseJitTypeAndSizeOfSIMDType(op3ClsHnd, immSimdSize);
+            *immSimdBaseType             = JitType2PreciseVarType(otherBaseJitType);
+            assert(otherBaseJitType == simdBaseJitType);
+        }
+        // For imm1 use default simd sizes.
```

Comment:
Should we be asserting here instead? It seems like there's a presumption about the value of `immSimdBaseType` and that the caller essentially shouldn't call this function for `immNumber == 1` for `InsertSelectedScalar`

Reply:
The code isn't clear here.

The `immSimdSize` and `immSimdBaseType` passed in are the default size and type for the intrinsic. The function can then choose to override them (as it does for `HW_Category_SIMDByIndexedElement` and imm2 of `InsertSelectedScalar`). For all other intrinsics (including imm1 of `InsertSelectedScalar`) it does nothing so that the default size and type can be used.

Updated the block comment so that it's clearer.